src/redisgears_python.c

```diff
@@ -65,145 +65,367 @@ static void* RedisGearsPy_PyCallbackDeserialize(FlatExecutionPlan* fep, Gears_Bu
+            return REDISMODULE_ERR;
+        }
+        fclose(f);
+
+        RedisGears_BWWriteString(bw, wheel);
+        RedisGears_BWWriteBuffer(bw, data, fsize);
 
-    session->requirementList = requirementsList;
+        RG_FREE(data);
+        RG_FREE(filePath);
+    }
+    return REDISMODULE_OK;
+}
 
+static void* PythonSessionCtx_ShellowCopy(void* arg){
+    PythonSessionCtx* session = arg;
+    ++session->refCount;
     return session;
 }
 
+static void PythonSessionCtx_FreeRequirementsList(PythonRequirementCtx** requirementsList){
+    for(size_t i = 0 ; i < array_len(requirementsList) ; ++i){
+        PythonRequirementCtx_Free(requirementsList[i]);
+    }
+    array_free(requirementsList);
+}
+
 static void PythonSessionCtx_Free(void* arg){
     PythonSessionCtx* session = arg;
     if(--session->refCount == 0){
+        // delete the session working dir
         Gears_dictDelete(SessionsDict, session->sessionId);
-        PythonSessionCtx_FreeRequirementsList(session->requirementList);
+        PythonSessionCtx_FreeRequirementsList(session->requirements);
         void* old = RedisGearsPy_Lock(NULL);
         Py_DECREF(session->globalsDict);
         RedisGearsPy_Unlock(old);
         RG_FREE(session);
     }
 }
 
+static PythonSessionCtx* PythonSessionCtx_Get(char* id){
+    return Gears_dictFetchValue(SessionsDict, id);
+}
+
+static PythonSessionCtx* PythonSessionCtx_CreateWithId(char* id, const char** requirementsList, size_t requirementsListLen){
+    // creating a new global dict for this session
+    void* old = RedisGearsPy_Lock(NULL);
+
+    PyObject* globalDict = PyDict_Copy(pyGlobals);
+
+    PythonSessionCtx* session = RG_ALLOC(sizeof(*session));
+    *session = (PythonSessionCtx){
+            .refCount = 1,
+            .globalsDict = globalDict,
+    };
+    SetId(NULL, session->sessionId, session->sessionIdStr, &CurrSessionId);
+
+    session->requirements = array_new(PythonRequirementCtx*, 10);
+
+    Gears_dictAdd(SessionsDict, session->sessionId, session);
+
+    for(size_t i = 0 ; i < requirementsListLen ; ++i){
+        PythonRequirementCtx* req = PythonRequirementCtx_Get(requirementsList[i]);
+
+        if(!req){
+            req = PythonRequirementCtx_Create(requirementsList[i]);
+
+            if(!req){
+                PythonSessionCtx_Free(session);
+                session = NULL;
+                break;
+            }
+
+            if(!PythonRequirementCtx_DownloadRequirement(req)){
+                PythonRequirementCtx_Free(req);
+                PythonSessionCtx_Free(session);
+                session = NULL;
+                break;
+            }
+
+            if(!PythonRequirementCtx_InstallRequirement(req)){
+                PythonRequirementCtx_Free(req);
+                PythonSessionCtx_Free(session);
+                session = NULL;
+                break;
+            }
+        }
+
+        session->requirements = array_append(session->requirements , req);
```

Comment:
session can be NULL here.

Reply:
How can it be NULL?

Reply:
You're right, it seemed to me it was outside the for.